retrystorage/retry_storage.go

```diff
@@ -0,0 +1,194 @@
+package retrystorage
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/microstorage"
+)
+
+const (
+	// defaultMaxAttempts is the default number of attempts of each storate
+	// operations before it finally fails.
+	defaultMaxAttempts = 3
+)
+
+type stoppingBackOff struct {
+	attempts    int
+	MaxAttempts int
+	Underlying  backoff.BackOff
+}
+
+func (s *stoppingBackOff) NextBackOff() time.Duration {
+	if s.attempts >= s.MaxAttempts {
+		return backoff.Stop
+	}
+	s.attempts++
+	return s.Underlying.NextBackOff()
+}
+
+func (s *stoppingBackOff) Reset() {
+	s.attempts = 0
+	s.Underlying.Reset()
+}
+
+type RetryStorageConfig struct {
+	// Dependencies.
+
+	Logger     micrologger.Logger
+	Underlying microstorage.Storage
```

Comment:
This concept is OK, but problem is that you can wrap `retrystorage.Storage` with `retrystorage.Storage`. And there is no easy way to work around it.